src/cudadev/DataFormats/SoALayout.h

```diff
@@ -0,0 +1,373 @@
+      }                                                                                                              \
+      CPP_TYPE::Scalar* BOOST_PP_CAT(addressOf_, NAME)() { return parent_.BOOST_PP_CAT(NAME, _); }                   \
+)
+// clang-format on
+#define _DEFINE_METADATA_MEMBERS(R, DATA, TYPE_NAME) _DEFINE_METADATA_MEMBERS_IMPL TYPE_NAME
+
+/**
+ * Member assignment for trivial constructor
+ */
+#define _DECLARE_MEMBER_TRIVIAL_CONSTRUCTION_IMPL(VALUE_TYPE, CPP_TYPE, NAME) \
+  _SWITCH_ON_TYPE(VALUE_TYPE,                       /* Scalar */              \
+                  (BOOST_PP_CAT(NAME, _)(nullptr)), /* Column */              \
+                  (BOOST_PP_CAT(NAME, _)(nullptr)), /* Eigen column */        \
+                  (BOOST_PP_CAT(NAME, _)(nullptr))(BOOST_PP_CAT(NAME, Stride_)(0)))
+
+#define _DECLARE_MEMBER_TRIVIAL_CONSTRUCTION(R, DATA, TYPE_NAME) \
+  BOOST_PP_EXPAND(_DECLARE_MEMBER_TRIVIAL_CONSTRUCTION_IMPL TYPE_NAME)
+/**
+ * Computation of the column or scalar pointer location in the memory layout (at SoA construction time)
+ */
+#define _ASSIGN_SOA_COLUMN_OR_SCALAR_IMPL(VALUE_TYPE, CPP_TYPE, NAME)                                                 \
+  _SWITCH_ON_TYPE(VALUE_TYPE, /* Scalar */                                                                            \
+                  BOOST_PP_CAT(NAME, _) = reinterpret_cast<CPP_TYPE*>(curMem);                                        \
+                  curMem += (((sizeof(CPP_TYPE) - 1) / byteAlignment) + 1) * byteAlignment;                           \
+                  , /* Column */                                                                                      \
+                  BOOST_PP_CAT(NAME, _) = reinterpret_cast<CPP_TYPE*>(curMem);                                        \
+                  curMem += (((nElements_ * sizeof(CPP_TYPE) - 1) / byteAlignment) + 1) * byteAlignment;              \
+                  , /* Eigen column */                                                                                \
+                  BOOST_PP_CAT(NAME, _) = reinterpret_cast<CPP_TYPE::Scalar*>(curMem);                                \
+                  curMem += (((nElements_ * sizeof(CPP_TYPE::Scalar) - 1) / byteAlignment) + 1) * byteAlignment *     \
+                            CPP_TYPE::RowsAtCompileTime * CPP_TYPE::ColsAtCompileTime;                                \
+                  BOOST_PP_CAT(NAME, Stride_) = (((nElements_ * sizeof(CPP_TYPE::Scalar) - 1) / byteAlignment) + 1) * \
+                                                byteAlignment / sizeof(CPP_TYPE::Scalar);)                            \
+  if constexpr (alignmentEnforcement == AlignmentEnforcement::Enforced)                                               \
+    if (reinterpret_cast<intptr_t>(BOOST_PP_CAT(NAME, _)) % byteAlignment)                                            \
+      throw std::out_of_range("In layout constructor: misaligned column: " #NAME);
```

Comment:
I did not find a perfect match in `std::exceptions` and felt this could be the closest (maybe `std::length_error` ? ). Otherwise, we could just create our own exceptions, but we mostly rely on `std::`  ones, don't we?

Reply:
I'd go with a generic `std::runtime_error`.